lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp

```diff
@@ -2192,17 +2192,40 @@ namespace Js
             return scriptContext->GetLibrary()->GetUndefined();
         }
 #else
-        int required = ValidateAndCanonicalizeTimeZone(tz->GetSz());
-        if (required > 0)
+        UErrorCode status = U_ZERO_ERROR;
+
+        // TODO(jahorto): Is this the list of timeZones that we want? How is this different from
+        // the other two enum values or ucal_openTimeZones?
+        UEnumeration *available = ucal_openTimeZoneIDEnumeration(UCAL_ZONE_TYPE_ANY, nullptr, nullptr, &status);
+        int availableLength = uenum_count(available, &status);
+        ICU_ASSERT(status, availableLength > 0);
+
+        charcount_t matchLen = 0;
+        UChar match[100];
+        for (int a = 0; a < availableLength; ++a)
         {
-            char16 *buffer = RecyclerNewArrayLeaf(scriptContext->GetRecycler(), char16, required);
-            ValidateAndCanonicalizeTimeZone(tz->GetSz(), buffer, required);
-            return JavascriptString::NewWithBuffer(buffer, required - 1, scriptContext);
+            int curLen = -1;
+            const UChar *cur = uenum_unext(available, &curLen, &status);
+            ICU_ASSERT(status, curLen > 0);
+            if (_wcsicmp(reinterpret_cast<const char16 *>(cur), tz->GetSz()) == 0)
+            {
+                ucal_getCanonicalTimeZoneID(cur, curLen, match, _countof(match), nullptr, &status);
+                ICU_ASSERT(status, true);
+                size_t len = wcslen(reinterpret_cast<const char16 *>(match));
+                AssertMsg(len < MaxCharCount, "Returned canonicalized timezone is far too long");
+                matchLen = (charcount_t)len;
+                break;
+            }
         }
-        else
+
+        uenum_close(available);
```

Comment:
>uenum_close [](start = 8, length = 11)

Do we have a strict guarantee that all of the above code can't throw? Does UEnumeration already do this for us in its destructor? If the answers are both no, then we need a wrapper class to make sure we tear down correctly.

Reply:
UEnumeration is a `void *` that is actually an `icu::StringEnumeration`, and `uenum_close()` just calls `delete stringEnum` in (almost?) all cases. So, it cant throw per destructor rules about throwing.

Reply:
My question wasn't about uenum_close throwing, it was about earlier exceptions making us fail to close. Sorry I was unclear.

---
In reply to: [169467337](https://github.com/Microsoft/ChakraCore/pull/4703#discussion_r169467337) [](ancestors = 169467337)